### ISSUE TITLE
Problem with inference for this['prop'] with conditional types

### ISSUE THREAD

**Issue by dfreeman** [microsoft/TypeScript#27014]
**TypeScript Version:**  typescript@3.1.0-dev.20180907
**Search Terms:** this type, conditional type

**Code**

```ts
type Wrapped<T> = { ___secret: T };
type Unwrap<T> = T extends Wrapped<infer U> ? U : T;

declare function set<T, K extends keyof T>(obj: T, key: K, value: Unwrap<T[K]>);

class Foo {
    prop: Wrapped<string>;

    method() {
        set(this, 'prop', 'hi'); // <-- type error
    }
}

set(new Foo(), 'prop', 'hi'); // <-- typechecks
```

**Expected behavior:**
Both calls to `set` (the one in the method with `this` and the external one with `new Foo()`) pass typechecking.

**Actual behavior:**
The external call passes, but the one inside `method` fails with `Argument of type '"hi"' is not assignable to parameter of type 'Unwrap<this["prop"]>'.`.

If I explicitly annotate `this: Foo` in the method signature or set the type parameters `set<Foo, 'prop'>(...)`, everything passes.

We've run into this in the Ember type declarations, and have had to [weaken the types for the time being](https://github.com/typed-ember/ember-cli-typescript/issues/286#) to work around it.

**Playground Link:** [available here](https://www.typescriptlang.org/play/#src=type%20Wrapped%3CT%3E%20%3D%20%7B%20___secret%3A%20T%20%7D%3B%0D%0Atype%20Unwrap%3CT%3E%20%3D%20T%20extends%20Wrapped%3Cinfer%20U%3E%20%3F%20U%20%3A%20T%3B%0D%0A%0D%0Adeclare%20function%20set%3CT%2C%20K%20extends%20keyof%20T%3E(obj%3A%20T%2C%20key%3A%20K%2C%20value%3A%20Unwrap%3CT%5BK%5D%3E)%3B%0D%0A%0D%0Aclass%20Foo%20%7B%0D%0A%20%20%20%20prop%3A%20Wrapped%3Cstring%3E%3B%0D%0A%0D%0A%20%20%20%20method()%20%7B%0D%0A%20%20%20%20%20%20%20%20set(this%2C%20'prop'%2C%20'hi')%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Aset(new%20Foo()%2C%20'prop'%2C%20'hi')%3B)

**Related Issues:** I wasn't able to find anything that seemed related to this issue.

**Comment by weswigham** [comment]
From what I know of conditional types, the root issue is that we don't perform any inferences to conditional types and `this` in the class body is still generic, so required successful inference to work correctly. #27012 may help here, but I'm not sure. It's certainly affected by the same area.

**Comment by dfreeman** [comment]
@weswigham thanks for the ping—I pulled and built master since that PR was merged, but still see the same error message with that sample code

**Comment by weswigham** [comment]
So, I've looked over it again, and while I do have a "fix" known, the desired behavior is unsound. Given
```ts
type Wrapped<T> = { ___secret: T };
type Unwrap<T> = T extends Wrapped<infer U> ? U : T;

declare function set<T, K extends keyof T>(obj: T, key: K, value: Unwrap<T[K]>): Unwrap<T[K]>;

class Foo {
    prop: Wrapped<string>;

    method() {
        set(this, 'prop', 'hi'); // <-- type error
    }
}
```
Consider subtype:
```ts
class Bar extends Foo {
  prop: Wrapped<"ohno">l
}
const x = new Bar();
x.method(); // We just set `prop` to `"hi"` even though this component only takes an `"ohno"`!
```

The "read" case of these types works fine:
```tstype Wrapped<T> = { ___secret: T };
type Unwrap<T> = T extends Wrapped<infer U> ? U : T;

declare function get<T, K extends keyof T>(obj: T, key: K): Unwrap<T[K]>;

class Foo {
    prop: Wrapped<string>;

    method() {
        return get(this, 'prop'); // <-- fine
    }
}

class Bar extends Foo {
    prop: Wrapped<"ohyeah">;
}

const a = get(new Foo(), 'prop'); // <-- also fine

const b = get(new Bar(), 'prop'); // <-- also fine
```

so @DanielRosenwasser as is, this is working as intended.

**Comment by dfreeman** [comment]
You're right that the specific example I gave was unsound with the subclassing you're describing, but as far as I can tell, that's not unique to anything with conditional types:

```ts
class Foo {
  prop: string;
  method() {
    this.prop = 'hi';
  }
}

class Bar extends Foo {
  prop: 'ohno';
}

const x = new Bar();
x.method(); // We just set `prop` to `"hi"` even though this component only takes an `"ohno"`!
```

**Comment by weswigham** [comment]
For partly historical reasons, we have a rule that allows the unsound writes for primitive clads member types since thinking about variance in simple code is a heavy burden (and tbh it predates us tracking variance), however generics and indexed accesses in general track variance strictly and disapprove of the unsound assignments.

**Comment by RyanCavanaugh** [comment]
This is working as intended. `this` in an instance method is polymorphic: the method can run on a subtype whose `prop` is narrower than `Foo["prop"]`.

For example, if `Bar extends Foo` declares `prop: Wrapped<"ohno">`, accepting `set(this, "prop", "hi")` in `Foo.method` would allow `new Bar().method()` to write `"hi"` where only `"ohno"` is valid. The argument therefore must be checked as `Unwrap<this["prop"]>`, not as `string`.

The direct assignment example has a historical permissive rule for primitive member writes. That does not extend to generic indexed accesses, which preserve the subtype-dependent property type. A value that varies with the receiver, such as `defaultProp: Unwrap<this["prop"]>` passed to `set`, is safe.

`tsc user-repro/index.ts` reports TS2345 at the call in `typescript@3.1.0-dev.20180907`; the same call is rejected by `typescript@7.1.0-dev.20260826.1`.
